### PR TITLE
[ICDS Dashboard] Dashboard usage report

### PR DESCRIPTION
**Pull request by challabeehyv**
##### SUMMARY
Implemented new dashboard usage report

##### FEATURE FLAG
yes

##### PRODUCT DESCRIPTION
This PR incudes both backend and ui integration of dashboard usage report

### CODE REVIEW

**Review by calellowitz — CHANGES_REQUESTED** [review]
A few high level comments. generally the idea seems right, just some of the details seem less optimal, especially for the national level users.

Additionally, is this tested for both xls and csv

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: plannig to add a default role 'N/A'. let me know if its okay

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: so i will bring only users under the logged in user location. or do i have to get the users in batches and add parallel processing?

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: plannig to add a default role 'N/A'. let me know if its okay

Comment (calellowitz) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: That should be fine

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (calellowitz) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: That is a good start, but is it not possible to do all the process once? Get a count for each user, then just look that number up in the user loop below. Instead of calculating that number each time.

Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: that is what happening there.The returned result set would be like this 
Screenshot from 2019-10-21 18-37-37. I am just looping to get the total count

Comment (calellowitz) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: RIght, but you do that looping for each user in the report. What if you do that looping only once. Something like
```
user_counts = defaultdict(int)
for r in records:
    user_counts[r['username']] += 1
```
then use user_counts in the loop below

Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: ohh understood

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/dashboard_usage.py]: please remove print statements

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/dashboard_usage.py]: location matrix looks it contains entire hierarchy list per row per awc. Why is this location_matrix needed?
The just descendants you can easily fine by  filtering AWCLocation on appropriate aggregation level.
which intern makes me thing `convert_rs_to_matrix` method is not needed.
M I missing something very obvious?

I m happy to discuss you would like more clarity.

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/dashboard_usage.py]: assuming you will be changing this logic based on Cals above comment.

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/dashboard_usage.py]: There are so many nested loops which will be very slow eventually and at higher level of hierarchy.

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/dashboard_usage.py]: If we already have user locations ids Why need yo query SQLLocation again because these locations will be sub locations of the logged in users whose info is already fetched above.

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Left a bunch of comments

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: We are converting the main location type into _id form and passing it to the method.

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: i did not get this..can you please elaborate

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: i dont think we will get hiercharchy structure with get_descendants. So constructed a matrix as we need ancestors and the hierarchy as well to contruct the excel data. Hope that is clear

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: The loops increase as there are more locations assigned to a single user which is not a general case i believe. The logic will cover national users as well

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: the assigned location ids from the couch user has only ids but not location types. So querying it again. As cal said, the will fetch the data and loop through this

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: added

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: added the fix

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: done

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: done

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: changed the flow..please check

**Comment by challabeehyv** [comment]
> Left a bunch of comments

changed the flow to eager fetch results and loop through them.please verify

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/dashboard_usage.py]: you would have to use the aggregation_level=5 else you will be lots of duplicate rows with entries as 'all' for lower locations.

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/dashboard_usage.py]: This is literally, fetching all rows for all the past months till nowfor that location and on every aggregate level which is not required and this could be very slow.

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/dashboard_usage.py]: could you brief about the intent of this function? its unclear to me whats actually happening here

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/dashboard_usage.py]: Row on this matrix will contain entire hierarchy right from state to awc?
then its not required. Because there is dashboard user which is at awc level.

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/dashboard_usage.py]: Is this block to fill 'All' to all locations lower than the location assigned to user?

Comment (Rohit25negi) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: and why do you thing user_location_row contains only state till block?

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Hey, Left a few logic correction comments. and I think there is good scope of cleaning this.

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: added

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: added

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: preparing a list of location_id vs is_launched and location_id vs location_type mapping

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: this is for excel data construction

**Review by calellowitz — APPROVED** [review]
I would really like to see that comment fixed, but won't block on it if we need to start QA now since it is purely a performance thing. If it is not fixed before merge, I would expect a followup within the week.

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: so should i not retreive supervisors as well ?

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: added the username filter.

File: custom/icds_reports/static/js/directives/download/download.directive.js
Comment (Rohit25negi) [inline custom/icds_reports/static/js/directives/download/download.directive.js]: Hey, This change has to be under feature flag. This looks this is open now.

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Last set of comments. But Looks good overall

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Very close to approval. Left one comment. 
Two more question part from that
1) did you tested it running manually localy
2) did you tested that feature flag being off this report is not visible on UI and only visible when feature flag is on for a partocular user

File: custom/icds_reports/static/js/directives/download/download.directive.js
Comment (challabeehyv) [reply custom/icds_reports/static/js/directives/download/download.directive.js]: added and verified

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: added national level user support

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: added the regex filter for usernames

File: custom/icds_reports/sqldata/exports/dashboard_usage.py
Comment (challabeehyv) [reply custom/icds_reports/sqldata/exports/dashboard_usage.py]: changed it

File: custom/icds_reports/static/icds_reports/js/spec/download.directive.spec.js
Comment (Rohit25negi) [inline custom/icds_reports/static/icds_reports/js/spec/download.directive.spec.js]: In this test cases I think FF is off so should not it be let 10? or M I missing somethign? @challabeehyv

**Review by Rohit25negi — APPROVED** [review]
Looks good. Thanks for all that hard work.

**Comment by challabeehyv** [comment]
> Very close to approval. Left one comment.
> Two more question part from that
> 
> 1. did you tested it running manually localy
> 2. did you tested that feature flag being off this report is not visible on UI and only visible when feature flag is on for a partocular user

i did both of them

File: custom/icds_reports/static/icds_reports/js/spec/download.directive.spec.js
Comment (challabeehyv) [reply custom/icds_reports/static/icds_reports/js/spec/download.directive.spec.js]: it has feature flag enabled. so should be 11

**Pull request merged**
